File: 212. Word Search II/solution.py
class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        """
        Variables used for complexity analysis:
        l: maximum length of a word
        w: number of words in the input list
        n: number of points on the board
        
        overall complexity: O(l*(w+n))
        """
        sol = []
        
        # edge cases - O(1)
        n = len(board)
        if not n:
            return sol
        m = len(board[0])
        if not m:
            return sol
        if not words:
            return sol
        
        # store words as a trie - O(w*l)
        size = n * m
        trie = {}
        for word in words:
            if len(word) > size:
                continue
            ptr = trie
            for w in word:
                if w not in ptr:
                    ptr[w] = {}
                ptr = ptr[w]
            ptr["#"] = "#"

        # dfs over each point on the board - O(n*l)
        sol = set()
        visited = set()
        for i in range(n):
            for j in range(m):
                if board[i][j] in trie:
                    self.dfs(board, i, j, visited, trie, sol, "")
                        
        return list(sol)
    
    # time complexity for dfs: O(l)
    def dfs(self, board, i, j, visited, trie, sol, prev):
        if "#" in trie:
            sol.add(prev)
            
        if i < 0 or j < 0 or i >= len(board) or j >= len(board[i]) or board[i][j] not in trie:
            return
        
        if (i, j) in visited:
            return
        
        visited.add((i, j))
        
        self.dfs(board, i-1, j, visited, trie[board[i][j]], sol, prev + board[i][j])
        self.dfs(board, i, j-1, visited, trie[board[i][j]], sol, prev + board[i][j])
        self.dfs(board, i+1, j, visited, trie[board[i][j]], sol, prev + board[i][j])
        self.dfs(board, i, j+1, visited, trie[board[i][j]], sol, prev + board[i][j]) 
        
        visited.remove((i, j))

File: 212. Word Search II/solution.java
class Solution {
    public List<String> findWords(char[][] board, String[] words) {
        List<String> sol = new ArrayList<String>();
        
        int m = board.length;
        if (m == 0) {
            return sol;
        }
        int n = board[0].length;
        if (n == 0) {
            return sol;
        }
        
        boolean found;
        boolean[][] visited = new boolean[board.length][board[0].length];
        for (String word: words) {
            found = false;
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    if (word.charAt(0) == board[i][j] && dfs(board, i, j, word, 0, visited)) {
                        sol.add(word);
                        found = true;
                        break;
                    }
                }
                if (found)
                    break;
            }
        }
        return sol;
    }
    
    public boolean dfs(char[][] board, int i, int j, String word, int k, boolean[][] visited) {
        if (k == word.length()) {
            return true;
        }
        
        if (i >= 0 && j >= 0 && i < board.length && j < board[0].length && 
            !visited[i][j] && board[i][j] == word.charAt(k)) {
            visited[i][j] = true;
            boolean res = dfs(board, i+1, j, word, k+1, visited) ||
            dfs(board, i, j+1, word, k+1, visited) ||
            dfs(board, i, j-1, word, k+1, visited) ||
            dfs(board, i-1, j, word, k+1, visited);
            visited[i][j] = false;
            return res;
        } else {
            return false;
        }
    }
}
